### PR TITLE
PortReport deprecated in favor of modern callbacks

### PR DESCRIPTION
**Pull request by aerydna**
the pr makes possible to pass a callable to everywhere the PortReport was used.

example 
`port.setReportCallback([ ](const yarp::os::PortInfo & info) { //do things });`
the major difference when passing a functor that has member variables that changes inside the operator() is that its required to wrapping it with std::ref()

### CODE REVIEW

**Review by drdanz — DISMISSED** [review]
@aerydna rebased and fixed several things.
It looks good to me now, but the release notes are still missing

**Comment by traversaro** [comment]
Just to understand, it would be helpful to provide an example of how an existing piece of code that uses a class that inherits from PortReport will change with this new pattern.

**Comment by pattacini** [comment]
@drdanz my point is that the class might be responsible for the callback plus some other handling; it very much depends on what type of computation is required.

Examples of this new usage that cover a wide range of cases may enrich the PR. 

Marking with 👎 anyone who doesn't agree with you is not a common attitude in our community @drdanz

File: src/libYARP_os/src/yarp/os/impl/PortCore.cpp [hunk 1]
```diff
@@ -1199,23 +1198,23 @@ void PortCore::describe(PortReport& reporter)
...
     }
 
     m_stateSemaphore.post();
 }
 
 
-void PortCore::setReportCallback(yarp::os::PortReport* reporter)
-{
+ void PortCore::setReportCallback(const std::function<void(const yarp::os::PortInfo&)>& reporter)
+ {
```
Comment (PeterBowman) [inline src/libYARP_os/src/yarp/os/impl/PortCore.cpp]: Leading spaces.

**Comment by drdanz** [comment]
> Marking with :-1: anyone who doesn't agree with you is not a common attitude in our community 

That was not personal, you proposed `Why don't we add up this new functionality on top without deprecating the good old method?` and I expressed my opinion against that proposal.

**Comment by drdanz** [comment]
> my point is that the class might be responsible for the callback plus some other handling; it very much depends on what type of computation is required.

This can still be done by passing a member of the class as callback

**Comment by traversaro** [comment]
> > my point is that the class might be responsible for the callback plus some other handling; it very much depends on what type of computation is required.
> 
> This can still be done by passing a member of the class as callback

Just to understand, this means that:
~~~
PortReportChild callbackFunctor;
port.setReportCallback(callbackFunctor);
~~~
becomes:
~~~
PortReportChild callbackFunctor;
port.setReportCallback(std::bind(&PortReportChild::report, &callbackFunctor, _1));
~~~
?

**Comment by pattacini** [comment]
@drdanz, of course, that wasn't personal, I was referring to a different context and attitude.

Anyway, my major concern was about the in-place definition of the callback as it was remarked from the example given in the PR description.

Lack of clarity in the PR description generates confusion and misunderstanding that we can easily avoid by providing examples – or at least more details – as already mentioned. This is because we don't have always time to review/peruse the PR content.

**Comment by drdanz** [comment]
Either `std::bind` or (better) using a lambda:

```c++
port.setReportCallback([](const yarp::os::PortInfo& info){ PortReportChild::report.callbackFunctor(info); });
```

**Comment by traversaro** [comment]
> Either `std::bind` or (better) using a lambda:
> 
> ```c++
> port.setReportCallback([](const yarp::os::PortInfo& info){ PortReportChild::report.callbackFunctor(info); });
> ```

I guess if you have an instance of `PortReportChild` , you should do as: 
> ```c++
> port.setReportCallback([&callbackFunctor](const yarp::os::PortInfo& info){ callbackFunctor.report.callbackFunctor(info); });
> ```

**Comment by drdanz** [comment]
> @drdanz, of course, that wasn't personal, I was referring to a different context and attitude.

Again, I have no idea what you are talking about. :-1: is just a reaction to a comment. If we have some misunderstanding, feel free to come talk to me.

I agree that some example might be useful, but nonetheless, this is a very basic C++11 callback approach that has been around for about 9 years. It's more or less the same that is used in almost any `std` algorithm, I don't think it requires a big explanation.

File: src/libYARP_os/src/yarp/os/impl/PortCore.cpp (hunk 1 above)
Comment (drdanz) [reply src/libYARP_os/src/yarp/os/impl/PortCore.cpp]: Fixed, thanks

**Comment by aerydna** [comment]
closing for obsolescence

**Pull request closed without merging**